Sources/CasePaths/EnumReflection.swift

```diff
@@ -98,6 +119,36 @@ public func extract<Root, Value>(_ embed: @escaping (Value) -> Root) -> (Root) -
 
 // MARK: - Private Helpers
 
+private func canPayloadType(_ payloadType: Any.Type, bePassedAndReturnedAs desiredType: Any.Type) -> Bool {
+  if payloadType == desiredType {
+    return true
+  }
+
+  // Consider this: `enum E { case c(l: Int) }`
+  //
+  // At the metadata level, c's associated value has type `(l: Int)`, which is a single-element tuple.
+  //
+  // But Swift doesn't support single-element tuples in Swift source code, so `CasePath<E, (l: Int)>` isn't allowed.
+  //
+  // The case path for `c` therefore has type `CasePath<E, Int>`.
+  //
+  // However, `(l: Int)` and `Int` have the same storage layout, so we can bit-cast one to the other safely. Therefore, if `payloadType` is a single-element tuple whose payload type is `desiredType`, I treat them as equivalent.
+
+  // TODO
+
+  // Consider these: `enum F { case d(a: Int, b: String) }`
+  //
+  // Certainly we should allow `CasePath<F, (a: Int, b: String)>` to match this.
+  //
+  // We would also like `CasePath<F, (Int, String)> to match this.
+  //
+  // So if `payloadType` is a tuple, and `desiredType` is a tuple with no labels, and the two types have identical elements types, I treat them as equivalent.
+
+  // TODO
```

Comment:
We explored handling this stuff in another branch, and particularly a couple more cases here: https://github.com/pointfreeco/swift-case-paths/blob/20d0a0668731d1f59dec45ea115e980e32feac47/Sources/CasePaths/EnumReflection.swift#L132-L141

Protocol conformances might be tough, but maybe there's a simpler way we didn't know about.

Reply:
I don't understand the “handle protocol conformances” case.

Consider

```
protocol P { }
extension Int: P { }
enum E {
    case p(P)
    case i(Int)
}
```

Then the only legal `embed` functions we can pass to `extract` are these:

- `E.p`, which takes a `P` (existential) argument, and the returned extractor's type is `(E) -> P?`
- `E.i`, which takes an `Int` argument, and the returned extractor's type is `(E) -> Int?`

In neither case does the extractor need to convert a `P` existential to an `Int` or vice versa. Am I overlooking something?

Reply:
In the above branch there is an assertion that fails if that `case let (fieldType as ExistentialMetadata...)` check is omitted:

```swift
    struct MyError: Equatable, Error {}

    XCTAssertEqual(
      .some(MyError()),
      (/Result<String, Error>.failure)
        .extract(from: .failure(MyError()))
    )
```

This is because the case path's `Value.self` is the `Error` existential, but the type we get from `swift_getTypeByMangledNameInContext` is `MyError`. I'm not sure if there's a simple means of making this `MyError is Error` test dynamically, which is why we have the more complicated metadata tests above.

Reply:
(You could argue that this kind of case path is rare and that Swift's type system limitations means we shouldn't bother supporting it, but I wanted to note that it _does_ currently work with reflection and the metadata code from your last PR.)

Reply:
Okay, I see what's going on.

Swift infers a case path type of `CasePath<Result<String, Error>, MyError>`. It has to, because `XCTAssertEqual` requires `Equatable` conformance, and `MyError` is `Equatable` but `Error` is not.

Without constraints, the type of `Result<String, Error>.failure` is `(Error) -> Result<String, Error>`. But here, Swift needs the type to be `(MyError) -> Result<String, Error>`.

Because `Error` is in a contravariant position, `(Error) -> Result<String, Error>` is a subtype of `(MyError) -> Result<String, Error>`. Swift implicitly converts the function to the supertype (by wrapping it in a “reabstraction thunk helper”) to make the expression type-check. How gracious.

So we have an expression that looks fine syntactically (you're apparently passing an enum case initializer, as required), but  where the `Value` type is not the associated value's type.

Yes, I think we should support this, but perhaps just by falling back to `Mirror` and a `#if DEBUG` warning.

Reply:
It turns out my last PR also suffers from this bug. I'm working on it.

Reply:
@mayoff Interesting, we probably didn't have good enough coverage on that. Did you cook up a failing test on `main`?

I updated this test and it still passes, though the case paths are pretty explicit in the types they return: https://github.com/pointfreeco/swift-case-paths/commit/1e224002a8a10ff8299dc0af72f82502150f2cc0

It's the end of the week, so energy is running a lil low, but it seems like even if your PR has this issue, I'm having a hard time thinking of a real world instance of it.

Reply:
Here's a failing test case:

https://github.com/mayoff/swift-case-paths/blob/41784e12a432daae9b4c23557b3796fc3af5db9a/Tests/CasePathsTests/CasePathsTests.swift#L361-L381

You'll also need the declarations related to `TestProtocol` from the top of the file.